src/client/Pipeline.cpp

```diff
@@ -241,18 +271,41 @@ void PipelineImpl::buildForAppendOrRecovery(bool recovery) {
              */
             if (stage != PIPELINE_SETUP_CREATE && stage != PIPELINE_CLOSE
                     && static_cast<int>(nodes.size()) < replication && canAddDatanode) {
-                if (!addDatanodeToPipeline(excludedNodes)) {
+                // Single data node case
+                bool added = false;
+                if (nodes.empty() && useRemoved) {
+                    if (storageID.length()) {
+                        LOG(INFO, "Pipeline: Adding back only datanode %s", removed.formatAddress().c_str());
+                        nodes.push_back(removed);
```

Comment:
Right, what you are missing is that the filesystem command updatePipeline is making RPC call to the server to change state there and it potentially modifies the block structure, so no that cannot be avoided.